### PR TITLE
Check based execution results

### PR DESCRIPTION
**Pull request by nelsonkopliku**
# Description

First iteration of a check based result visualization of executions.

<details>
  <summary>Some checks selected</summary>
  
check-based-results-pt1

</details>

<details>
  <summary>All checks selected</summary>
  
check-based-results-many

</details>

<details>
  <summary>One target timing out</summary>
  
one-agent-timeout

</details>

<details>
  <summary>All targets timing out</summary>
  
all-agents-timeout

</details>

This is an already a big enough PR, so the following has been deferred
- Add error handling on catalog/executionData loading failure
- Bind filtering by result
- Open remediation suggestion modal when clicking on check id 
- UI improvements (spacing, hovering, etc...)
- investigate/improve edge cases (ie timeout+expect_same)

## How was this tested?

Existing automated tests + newly added tests.

### CODE REVIEW

**Review by fabriziosestito — COMMENTED** [review]
@nelsonkopliku could you rebase on top of `new-checks-ui` branch and remove the "new" components?

**Review by abravosuse — APPROVED** [review]
Code aside, it looks great to me!

File: assets/js/components/NewChecksResults/ClusterChecksResultsPage.jsx [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import { getLastExecution } from '@state/selectors/lastExecutions';
+import { getCluster, getClusterHostNames } from '@state/selectors/cluster';
+import { updateCatalog } from '@state/actions/catalog';
+import { updateLastExecution } from '@state/actions/lastExecutions';
+import { RUNNING_EXECUTION_STATE } from '@state/lastExecutions';
+import ExecutionHeader from './ExecutionHeader';
+import ExecutionResults from './ExecutionResults';
+import { getCheckResults } from '../ExecutionResults/checksUtils';
+
+function ClusterChecksResultsPage() {
```
Comment (arbulu89) [inline assets/js/components/NewChecksResults/ClusterChecksResultsPage.jsx:14]: Why not `ExecutionResultsPage` as it was before?
We are bouncing around this name all the time hehe

File: assets/js/components/NewChecksResults/ExecutionHeader.jsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import React from 'react';
```
Comment (arbulu89) [inline assets/js/components/NewChecksResults/ExecutionHeader.jsx:1]: This component is something we use in other places, right?
Maybe making the filter optional we could have it as reusable component.
Adding the option to change the `span` text and changing the name to something more generic.
Something to have in our radar.

File: assets/js/components/NewChecksResults/ExecutionResults.jsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import { ExecutionIcon } from '@components/ExecutionResults/ExecutionIcon';
+import { getCheckDescription } from '@components/ExecutionResults/checksUtils';
+import CheckResultOutline from './CheckResultOutline';
+
+const resultsTableConfig = {
+  usePadding: false,
+  columns: [
+    {
+      title: 'Id',
+      key: 'checkId',
```
Comment (arbulu89) [inline assets/js/components/NewChecksResults/ExecutionResults.jsx:14]: `checkId` to `checkID`

File: assets/js/components/NewChecksResults/ExecutionResults.jsx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  });
+
+function ExecutionResults({
+  clusterName,
+  executionState,
+  checksResults,
+  catalog,
+  hostnames,
+}) {
+  const tableData = checksResults.map(
```
Comment (arbulu89) [inline assets/js/components/NewChecksResults/ExecutionResults.jsx:69]: Maybe we should have some internal mapper to map the result coming from wanda to our internal object.
Otherwise, it looks like we are pretty much forced to do operations in many places, like adding this hostnames to targets and use variables with `_`.

File: assets/js/components/NewChecksResults/ExecutionResults.test.jsx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+  addCriticalExpectation,
+  addPassingExpectation,
+  agentCheckResultFactory,
+  checkResultFactory,
+  withEmptyExpectations,
+} from '@lib/test-utils/factories';
+import '@testing-library/jest-dom';
+import ExecutionResults from './ExecutionResults';
+
+const buildEmptyCheckResult = (checkId, agents, result) => {
```
Comment (arbulu89) [inline assets/js/components/NewChecksResults/ExecutionResults.test.jsx:19]: `checkdId` to `checkID`

File: assets/js/state/selectors/cluster.js [hunk 6]
```diff
@@ -10,6 +10,13 @@ export const getClusterHostIDs =
       .filter((host) => host.cluster_id === clusterID)
       .map(({ id: hostID }) => hostID);
 
+export const getClusterHostNames =
```
Comment (arbulu89) [inline assets/js/state/selectors/cluster.js:13]: The name is a bit misleading. Maybe we could have `getClusterHosts` and return all the fields?

File: assets/js/components/NewChecksResults/CheckResultOutline.jsx [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+const isExpect = ({ type }) => type === 'expect';
+const isExpectSame = ({ type }) => type === 'expect_same';
+
+function CheckResultOutline({
+  checkId,
+  agentsCheckResults,
+  expectationResults,
+  clusterName,
+}) {
+  const expectExpectationsCount = expectationResults.filter(isExpect).length;
```
Comment (arbulu89) [inline assets/js/components/NewChecksResults/CheckResultOutline.jsx:21]: We could have this `expectExpectationsCount` inside the next `map` function I guess

File: assets/js/components/NewChecksResults/CheckResultOutline.jsx [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  const expectResults = agentsCheckResults.map((agentCheckResult) => {
+    const { hostname, expectation_evaluations } = agentCheckResult;
+
+    const isError = !!agentCheckResult?.type;
```
Comment (arbulu89) [inline assets/js/components/NewChecksResults/CheckResultOutline.jsx:25]: Maybe we could wrap this tiny `!!agentCheckResult?.type;` somehow as `isAgentCheckError` or similar.
At least, to me personally, it would help to know that we are dealing with a `AgentCheckError` object type and not a `AgentCheckResult` object

**Review by arbulu89 — COMMENTED** [review]
@nelsonkopliku I really liked the work!
I'm fine with creating a new route by now (even though it will be merged to a dedicated PR). This way we can do a final "cleanup" PR to remove all leftovers.

I'm just adding a first comment so we can discuss on them.

Please track all the mentioned "missing" things in our internal ticket, so we are aware of them. Even though I guess you are already working on them while we review hehe

Some generic comments:
- I would name the main component `NewExecutionResults`. To not bounce around on names.
- The `expectations` column is missing in the main rows.
- Already mentioned in some comments. `checkId` -> `checkID`

**Comment by fabriziosestito** [comment]
> I'm fine with creating a new route by now (even though it will be merged to a dedicated PR). This way we can do a final "cleanup" PR to remove all leftovers.

@nelsonkopliku @arbulu89 
I would avoid this honestly, we have a new branch and we don't need to do it as it creates extra work prone to error to clean-up leftovers.

File: assets/js/components/NewChecksResults/ExecutionResults.test.jsx [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    agents_check_results: agents.map((agentId) =>
+      agentCheckResultFactory.build({
+        agent_id: agentId,
+      })
+    ),
+    result,
+  });
+
+  return withEmptyExpectations(checkResult);
+};
```
Comment (dottorblaster) [inline assets/js/components/NewChecksResults/ExecutionResults.test.jsx:31]: Couldn't you just have a factory for this?

**Review by dottorblaster — CHANGES_REQUESTED** [review]
Remove the `New` component as we decided some days ago then have a look at the comments, then we can merge :+1:

File: assets/js/components/NewChecksResults/CheckResultOutline.jsx (hunk 7 above)
Comment (nelsonkopliku) [reply assets/js/components/NewChecksResults/CheckResultOutline.jsx]: Not really I guess, we don't need that expression evaluated for each mapped element :thinking: 
What am I missing?

File: assets/js/components/NewChecksResults/CheckResultOutline.jsx (hunk 8 above)
Comment (nelsonkopliku) [reply assets/js/components/NewChecksResults/CheckResultOutline.jsx]: changed to `const isAgentCheckError = !!agentCheckResult?.type;`

File: assets/js/components/NewChecksResults/ClusterChecksResultsPage.jsx (hunk 1 above)
Comment (nelsonkopliku) [reply assets/js/components/NewChecksResults/ClusterChecksResultsPage.jsx]: Having removed the `New` thingy this does not apply anymore.

However something to take into account as we evolve is that checks results would be scoped byCluster, bySAPSystem, byWhatever. But let's see how that evolves, for now we should be good.

File: assets/js/components/NewChecksResults/ExecutionHeader.jsx (hunk 2 above)
Comment (nelsonkopliku) [reply assets/js/components/NewChecksResults/ExecutionHeader.jsx]: Not yet used in other places, however it could be enhanced to make it reusable in the cluster details where we have a similar visualization IIRC.

File: assets/js/components/NewChecksResults/ExecutionResults.jsx (hunk 4 above)
Comment (nelsonkopliku) [reply assets/js/components/NewChecksResults/ExecutionResults.jsx]: Well, totally fair point. I guess that intervention needs some extra thinking and restructuring of the app.
Food for our frontend guild :smile:

File: assets/js/state/selectors/cluster.js (hunk 6 above)
Comment (nelsonkopliku) [reply assets/js/state/selectors/cluster.js:13]: I followed the train of thoughts of the previous `getClusterHostIDs`, which might need to be revisited as well based on this suggestion.

If it's not a hard requirement changing now, I'd keep it like this and iterate over at a later stage :sweat_smile:

File: assets/js/state/selectors/cluster.test.js [hunk 10]
```diff
@@ -24,6 +28,41 @@ describe('Cluster selector', () => {
     expect(getClusterHostIDs('cluster1')(state)).toEqual(['id1', 'id2']);
   });
 
+  it('should return the cluster hosts names', () => {
+    const state = {
+      hostsList: {
+        hosts: [
```
Comment (EMaksy) [inline assets/js/state/selectors/cluster.test.js:34]: Hey what about using the hostFactory for that?

Something like this ? 
```
hosts: [
      hostFactory.build({ params: { id: 'id1', cluster_id: 'cluster1', hostname: 'hostname-1' } }),
      hostFactory.build({ params: { id: 'id2', cluster_id: 'cluster1', hostname: 'hostname-2' } }),
      hostFactory.build({ params: { id: 'id3', cluster_id: 'cluster2', hostname: 'hostname-3' } }),
    ]
```
    
You can find the the host factory here:
`assets/js/lib/test-utils/factories/hosts.js`

Comment (nelsonkopliku) [reply assets/js/state/selectors/cluster.test.js:34]: Sure, thanks!

**Comment by nelsonkopliku** [comment]
@trento-project/developers addressed all the comments.
I'll keep iterating on this.

**Comment by jagabomb** [comment]
@nelsonkopliku I only realised now that the screenshots are in the accordions/dropdowns... Anyway judging by the screenshots I think its looking great. I have a couple of suggestions. 

Firstly, I'm not sure if the table header is the same colour as the background of Trento. In the design, I prefer to highlight the table header slightly by giving it a background of Gray-50. If it is there then just ignore the comment as I am judging by the screenshots. 

Secondly, I would like to retain the hover effect of a green highlight when hover over a table row. We have this interaction currently on demo which can be seen here: http://demo.trento-project.io/clusters/7965f822-0254-5858-abca-f6e8b4c27714/executions/last

**Review by dottorblaster — APPROVED** [review]
Besides @jagabomb's suggestions, LGTM

**Review by rtorrero — APPROVED** [review]
Hey Nelson, great work! code looks good!

PS: does it make sense to make the expectation details row not clickable for now and defer that for the next PRs?

**Review by EMaksy — APPROVED** [review]
Looks great!
